File: src/solutions/scala/N20N32Parentheses.scala
package solutions.scala

import scala.annotation.tailrec

object N20N32Parentheses {

  def validParentheses(s: String): Boolean = {
    val opening = Set('(', '[', '{')
    val equivalences = Map(')' -> '(', ']' -> '[', '}' -> '{')

    @tailrec
    def validParentheses(l: List[Char], stack: List[Char]): Boolean = l match {
      case Nil => stack.isEmpty
      case h :: t =>
        if (opening.contains(h)) validParentheses(t, h :: stack)
        else if (stack.nonEmpty && equivalences(h) == stack.head) validParentheses(t, stack.tail)
        else false
    }

    validParentheses(s.toList, Nil)
  }

  def longestValidParentheses(s: String): Int = {
    @tailrec
    def longestValidParentheses(l: List[Char], stack: List[Int], n: Int, maxLength: Int): Int = l match {
      case Nil => maxLength
      case h :: t =>
        if (h == '(') longestValidParentheses(t, n :: stack, n + 1, maxLength)
        else {
          if (stack.tail.isEmpty) longestValidParentheses(t, n :: stack.tail, n + 1, maxLength)
          else longestValidParentheses(t, stack.tail, n + 1, Math.max(maxLength, n - stack.tail.head))
        }
    }

    longestValidParentheses(s.toList, List(-1), 0, 0)
  }

  def simpleBrackets(s: String): Boolean = {
    def bracketValue(b: Char): Int = b match {
      case '(' => 1
      case ')' => -1
      case _ => 0
    }

    @tailrec
    def simpleBrackets(l: List[Char], n: Int): Boolean = (l, n) match {
      case (_, _) if n < 0 => false
      case (Nil, nn) => nn == 0
      case (h :: t, nn) => simpleBrackets(t, nn + bracketValue(h))
    }

    simpleBrackets(s.toList, 0)
  }
}

File: src/solutions/java/N20N32Parentheses.java
package solutions.java;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Map;
import java.util.Set;

public class N20N32Parentheses {

    private static final Set<Character> OPENING = Set.of('(', '[', '{');
    private static final Map<Character, Character> EQUIVALENCES = Map.of(')', '(', ']', '[', '}', '{');

    public static boolean isValid(String s) {
        Deque<Character> stack = new ArrayDeque<>(s.length() / 2);
        for (char c : s.toCharArray()) {
            if (OPENING.contains(c)) {
                stack.push(c);
            } else {
                if (EQUIVALENCES.get(c).equals(stack.peek())) {
                    stack.pop();
                } else {
                    return false;
                }
            }
        }
        return stack.isEmpty();
    }

    public static int longestValidParentheses(String s) {
        int maxLength = 0;
        int left = 0;
        int right = 0;
        for (Character c : s.toCharArray()) {
            if (c.equals('(')) {
                left++;
            } else {
                right++;
            }
            if (left == right) {
                maxLength = Math.max(maxLength, 2 * right);
            } else if (right >= left) {
                left = 0;
                right = 0;
            }
        }
        left = 0;
        right = 0;
        for (int i = s.length() - 1; i >= 0; i--) {
            if (s.charAt(i) == '(') {
                left++;
            } else {
                right++;
            }
            if (left == right) {
                maxLength = Math.max(maxLength, 2 * left);
            } else if (left >= right) {
                left = 0;
                right = 0;
            }
        }
        return maxLength;
    }

    public static int longestValidParenthesesStack(String s) {
        int maxLength = 0;
        Deque<Integer> stack = new ArrayDeque<>();
        stack.push(-1);
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') {
                stack.push(i);
            } else {
                stack.pop();
                if (stack.isEmpty()) {
                    stack.push(i);
                } else {
                    maxLength = Math.max(maxLength, i - stack.peek());
                }
            }
        }
        return maxLength;
    }

    public static void main(String[] args) {
        System.out.println(N20N32Parentheses.isValid("({({})([])})"));
        System.out.println(N20N32Parentheses.isValid("]"));
        System.out.println(N20N32Parentheses.longestValidParentheses("())(())"));
        System.out.println(N20N32Parentheses.longestValidParentheses(")()())"));
        System.out.println(N20N32Parentheses.longestValidParenthesesStack("())(())"));
        System.out.println(N20N32Parentheses.longestValidParenthesesStack(")()())"));
    }
}
